modules/apps/frontend-token/frontend-token-definition-impl/src/main/java/com/liferay/frontend/token/definition/internal/FrontendTokenDefinitionManager.java

```diff
@@ -0,0 +1,189 @@
+			themeIdFrontendTokenDefinitions) {
+
+		_jsonFactory = jsonFactory;
+		_themeIdFrontendTokenDefinitionImplsDCLSingleton =
+			themeIdFrontendTokenDefinitionImplsDCLSingleton;
+		_allCompaniesFrontendTokenDefinitions =
+			allCompaniesFrontendTokenDefinitions;
+
+		_themeIdFrontendTokenDefinitionImpls = themeIdFrontendTokenDefinitions;
+	}
+
+	public void addFrontendTokenDefinition(
+		long companyId, String externalReferenceCode,
+		String frontendTokenDefinitionAsJsonString) {
+
+		_addFrontendTokenDefinitionFromClientExtension(
+			companyId, externalReferenceCode,
+			frontendTokenDefinitionAsJsonString,
+			ResourceBundleLoaderUtil.getPortalResourceBundleLoader());
+	}
+
+	public FrontendTokenDefinitionImpl addFrontendTokenDefinition(
+		String themeId, ResourceBundleLoader resourceBundleLoader,
+		String frontendTokenDefinitionAsJsonString) {
+
+		return _addFrontendTokenDefinitionFromWorkspace(
+			frontendTokenDefinitionAsJsonString, resourceBundleLoader, themeId);
+	}
+
+	public FrontendTokenDefinition getFrontendTokenDefinition(
+		long companyId, String externalReferenceCode) {
+
+		return _getCompanyFrontendTokenDefinitions(
+			companyId
+		).get(
+			externalReferenceCode
+		);
+	}
+
+	public Map<String, FrontendTokenDefinitionImpl>
+		getThemeIdFrontendTokenDefinitionImpls() {
+
+		return _themeIdFrontendTokenDefinitionImpls;
+	}
+
+	public DCLSingleton<Map<String, FrontendTokenDefinitionImpl>>
+		getThemeIdFrontendTokenDefinitionImplsDCLSingleton() {
+
+		return _themeIdFrontendTokenDefinitionImplsDCLSingleton;
+	}
+
+	public void removeFrontendTokenDefinition(Long companyId, String key) {
+		if (Objects.nonNull(companyId)) {
+			_getCompanyFrontendTokenDefinitions(
+				companyId
+			).remove(
+				key
+			);
+		}
+		else {
+			_themeIdFrontendTokenDefinitionImpls.remove(key);
+		}
+	}
+
+	public void removeFrontendTokenDefinition(String key) {
+		removeFrontendTokenDefinition(null, key);
+	}
+
+	private void _addFrontendTokenDefinitionFromClientExtension(
+		Long companyId, String externalReferenceCode,
+		String frontendTokenDefinitionAsJsonString,
+		ResourceBundleLoader resourceBundleLoader) {
+
+		try {
+			Objects.requireNonNull(frontendTokenDefinitionAsJsonString);
+			Objects.requireNonNull(companyId);
+			Objects.requireNonNull(externalReferenceCode);
+
+			FrontendTokenDefinitionImpl frontendTokenDefinitionImpl =
+				_newFrontendTokenDefinitionImpl(
+					frontendTokenDefinitionAsJsonString, resourceBundleLoader);
+
+			if (frontendTokenDefinitionImpl != null) {
+				_getCompanyFrontendTokenDefinitions(
+					companyId
+				).put(
+					externalReferenceCode, frontendTokenDefinitionImpl
+				);
+			}
+		}
+		catch (JSONException jsonException) {
+			throw new RuntimeException(jsonException);
+		}
+	}
+
+	private FrontendTokenDefinitionImpl
+		_addFrontendTokenDefinitionFromWorkspace(
+			String frontendTokenDefinitionAsJsonString,
+			ResourceBundleLoader resourceBundleLoader, String themeId) {
```

Comment:
**themeId** is just used on workspace flow. It comes from the bundle file _liferay-look-and-feel.xml_.
CET flow uses _external-reference-code_ and _companyId_

Reply:
Okay, I think I'm understanding now. I don't think this should be called workspace flow. This is a frontendTokenDefinition coming from a theme rather than one coming from the a themeCSS CX. Those are currently the two flows: `theme` and `themeCSS CX`. 

From a workspace we have the possibility to create a theme or a themeCSS CX so that's probably not a good way to handle this differentiation. 

Does that make sense? @thiago-buarqque @algobob

Reply:
It does. We updated it to `_addBundleFrontendTokenDefinition` and the CET one to `_addThemeCSSCETFrontendTokenDefinition`. Once we code the FrontendTokenDefinition support for workspace CXs we'll also call `_addThemeCSSCETFrontendTokenDefinition`.

Reply:
That's better, but bundle still sounds too generic... nearly everything in Liferay that can be deployed get's called a bundle. I think `_addThemeFrontendTokenDefinition` would be more clear. Up until this PR is merged the only way to provide a frontendTokenDefinition to Liferay is to create a custom theme. So theme is probably the best word.

Reply:
We thought "theme" could reference ThemeCSS CX or Theme bundle. Not making it clear which one is being referenced. Do you think this is valid? If so, would `_addThemeBundleFrontendTokenDefinition` be a better name?

Reply:
"[theme](https://learn.liferay.com/w/dxp/site-building/site-appearance/themes/introduction-to-themes?p_l_back_url=%2Fsearch%3Fq%3Dtheme&highlight=theme)" has a lot of history in Liferay. It really only refers to one thing: a theme module that includes css, js, templates, etc. It's distinct from a themeCSS client extension. A themeCSS client extension let's you replace the CSS from a theme. So I think just "theme" is better than "themeBundle".